### PR TITLE
Fix falling node infinite falling on overhigh collision box, and rotation of facedir nodes

### PR DESCRIPTION
**Pull request by Wuzzy2**
This PR fixes 2 bugs in falling nodes:

1. When a falling node falls on a node with an overhigh collision box (i.e. upper collision box Y > 1.0), the falling node stops (because it collides), then it teleports a short distance upwards, then falls again, and stops again at the overhigh collisionbox, etc. In other words, the falling node is stuck in an infinite loop. This is really bad.
    I use this as a fix: When the falling node suddenly stops (i.e. velocity = 0), it drops as item(s). Because if a falling node hits an overhigh collision box, there is no clean way to actually place the node.

2. Partial fix of #5324. If the node uses facedir, wallmounted, colorfacedir, or colorwallmounted, the yaw of the entity is set properly. Limitation: If the axis of the node is rotated as well, the rotation is still wrong. The full fix is not possible with current entities as entities can't pitch or roll.

No. 1 was tested successfully with an overhigh Minetest Game fence.
No. 2 was tested successfully with a Minetest Game chest, but with `falling_node=1`. And some other nodes.

### CODE REVIEW

File: builtin/game/falling.lua [hunk 1]
```diff
@@ -27,12 +27,21 @@ core.register_entity(":__builtin:falling_node", {
 			is_visible = true,
 			textures = {node.name},
 		})
+		local def = core.registered_nodes[node.name]
+		if node.param2 ~= 0 then
```
Comment (SmallJoker) [inline builtin/game/falling.lua]: `if def and node.param2 ~= 0`
For the special case when falling node entities are saved, the mod gets disabled and this node becomes an unknown one.

**Comment by paramat** [comment]
So in your PR if a falling node entity lands on a overhigh node it drops as an item.
However, if a falling node entity drops on a slab, which is an equivalent situation of a collisionbox top not at a cube border, it becomes a hovering node:

screenshot_20180129_155526

This also happens when you place a node on a slab.

> if a falling node hits an overhigh collision box, there is no clean way to actually place the node.

So i think it's better to place the node as in the screenshot for consistency.

File: builtin/game/falling.lua (hunk 1 above)
Comment (Wuzzy2) [reply builtin/game/falling.lua]: And this is is a problem … how?
Is the node table in a weird format when this happens?

**Comment by Wuzzy2** [comment]
I think every dev agrees that a node “floating” on top of a slab is NOT a bug, but intended behaviour. This was already this way before this PR. The reason is that nodes should obey the grid, even if it sometimes leads to somewhat odd situations.

Also, slabs don't have an overhigh collisionbox, so this comparison doesn't make sense.

But damn, `ignore` is indeed a problem. How can I test this? I can't just place `ignore`.

The interesting thing about overhigh collisionboxes is that … well … they're overhigh. If you really want to place the node, then where? You first need to find out the position of the overhigh node AND probably also `get` it.

Do you have a better idea on how to fix infinite loops?
What are the other reasons for falling nodes to suddenly stop?

**Comment by paramat** [comment]
> Also, slabs don't have an overhigh collisionbox, so this comparison doesn't make sense.

I mean that a 1.5 high node with a 1.5 high collisionbox is visually identical to a slab on a node, so in both cases i think falling nodes floating above with a gap is fine and consistent.

**Comment by Wuzzy2** [comment]
> so in both cases i think falling nodes floating above with a gap is fine and consistent.

Except the falling node is not *floating*, but falling in an infinite loop. This looks horrible and wastes CPU time for no reason.

In MineClone 2 I found a “solution” which works like this: When velocity reaches 0, it checks 2 nodes below. If it finds a fence or wall (collisionbox height = 1.5), the entity is removed and and the actual node is placed 1 unit above the fence or wall node.
It's not a good solution as its only works as the only overhigh collisionboxes are fences and walls.
Also, the falling on a fence is not really smooth anymore. At least it fixes the bug in stock MineClone 2, but I'm not happy with this.


Or maybe we should just forbid using overhigh collisionboxes altogether as they cause so many problems? In ths idea, any overhigh collisionbox coulld be implemented by adding an additional invisible node for the overhigh part.

**Comment by paramat** [comment]
> Except the falling node is not floating, but falling in an infinite loop.

I know, i'm suggesting placing it as a node at the suggested position. However i realise this is not easy to do.
> which works like this: When velocity reaches 0, it checks 2 nodes below. If it finds a fence or wall (collisionbox height = 1.5), the entity is removed and and the actual node is placed 1 unit above the fence or wall node.

Instead of checking for specific nodes maybe check that the position of the top of the collisionbox ends in a .5?
> Or maybe we should just forbid using overhigh collisionboxes altogether

Nooo.

File: builtin/game/falling.lua (hunk 1 above)
Comment (SmallJoker) [reply builtin/game/falling.lua]: Yes, `def` is nil if it's an known node.

Comment (paramat) [reply builtin/game/falling.lua]: ^ unknown?

**Comment by paramat** [comment]
Part 1 needs a rethink.
Part 2 only fixes for nodes rotated around the y axis, 4 out of 24 rotations, so i'm not sure it's worth doing. Also most falling nodes are cubes not nodeboxes so the bug is not often seen.

**Comment by bigfoot547** [comment]
Yeah, you can place ignore :p

**Comment by bigfoot547** [comment]
Well... This is awkward, I seem to have lost the piece of code I modified to allow ignore to be placed... I'll work on this later when I'm not on some crappy old iPad ;)

**Comment by paramat** [comment]
> I use this as a fix: When the falling node suddenly stops (i.e. velocity = 0), it drops as item(s).

> Node entities stop for other reasons, such as hitting the border of an unloaded mapblock,

So will not work, and no progress, will reopen on progress.

**Pull request closed without merging**